start-training.py

```diff
@@ -18,252 +18,296 @@
+                config=base_config)
+            # Note that wandb config can contain different and/or new keys that
+            # aren't in the command-line arguments. This is especially true for
+            # wandb sweeps.
+            config.update(wandb.config._items)
+
+            # Save the environment type to the wandb summary data.
+            # This allows env_type show up in the benchmark table.
+            wandb.run.summary['env_type'] = config['env_type']
+
+            if launch_ingredients.job_name is None:
+                launch_ingredients.job_name = wandb.run.name
+                launch_ingredients.data_dir = os.path.join(
+                    launch_ingredients.safety_dir, 'data', time.strftime("%Y-%m-%d-") + wandb.run.id)
+
+            logging_setup.save_code_to_wandb()
+    else:
+        wandb = None
+        config.update(base_config)
+
+
+def setup_from_args(args):
+
+    # launch_ingredients is a place to store variables that are derived from
+    # the config but not part of it, and that are the result of our setup
+    # process
+    li = argparse.Namespace()  # aka launch_ingredients
+
+    if args.seed is None:
+        # Make sure the seed can be represented by floating point exactly.
+        # This is just because we may want to pass it over the web, and javascript
+        # doesn't have 64 bit integers.
+        args.seed = np.random.randint(2**53)
+
+    assert args.wandb or args.data_dir or args.run_type == 'inspect', (
+        "Either a data directory must be set or the wandb flag must be set. "
+        "If wandb is set but there is no data directory, then a run name will be "
+        "picked automatically.")
+
+    if args.ensure_gpu:
+        assert torch.cuda.is_available(), "CUDA support requested but not available!"
+
+    li.safety_dir = os.path.realpath(os.path.dirname(__file__))
+    sys.path.insert(1, li.safety_dir)  # ensure current directory is on the path
+    build_and_import_modules(li.safety_dir)
+
+    # Check to see if the data directory is already in use
+
+    if args.data_dir is not None:
+        li.data_dir = os.path.realpath(args.data_dir)
+        li.job_name = os.path.basename(li.data_dir)
+        if os.path.exists(li.data_dir) and args.run_type == 'train':
+            print("The directory '%s' already exists." % li.data_dir)
+            print("Would you like to overwrite the old data, append to it, or abort?")
+            response = 'overwrite' if li.job_name.startswith('tmp') else None
+            while response not in ('overwrite', 'append', 'abort'):
+                response = input("(overwrite / append / abort) > ")
+            if response == 'overwrite':
+                print("Overwriting old data.")
+                shutil.rmtree(li.data_dir)
+            elif response == 'abort':
+                print("Aborting.")
+                exit()
+    else:
+        li.data_dir = li.job_name = None
+
+    setup_config_and_wandb(args, li)
```

Comment:
One of the things I liked about doing it the previous way, without functions, was that all of the code in the start-training script was executed strictly in order. Now, when reading the code, you kind of have to read it out of order. For instance, when reading `setup_config_and_wandb` by itself, it's not obvious what the `launch_ingredients` object is and you basically need to skip down to this function to find out.